modules/transactions/src/main/java/org/apache/ignite/internal/tx/storage/state/rocksdb/TxStateRocksDbStorage.java

```diff
@@ -392,64 +259,225 @@ public long persistedIndex() {
         }
     }
 
-    /** {@inheritDoc} */
-    @Override public void destroy() {
-        try (Options options = new Options()) {
-            close();
+    /**
+     * Reads the value of {@link #lastAppliedIndex} from the storage.
+     *
+     * @param readOptions Read options to be used for reading.
+     * @return The value of last applied index.
+     */
+    private long readLastAppliedIndex(ReadOptions readOptions) {
+        byte[] appliedIndexBytes;
 
-            RocksDB.destroyDB(dbPath.toString(), options);
-        } catch (Exception e) {
-            throw new IgniteInternalException(TX_STATE_STORAGE_DESTROY_ERR, "Failed to destroy the transaction state storage", e);
+        try {
+            appliedIndexBytes = db.get(readOptions, lastAppliedIndexKey);
+        } catch (RocksDBException e) {
+            throw new IgniteInternalException(
+                TX_STATE_STORAGE_ERR,
+                "Failed to read applied index value from transaction state storage, partition " + partitionId
+                    + " of table " + tableStorage.configuration().value().name(),
+                e
+            );
         }
+
+        return appliedIndexBytes == null ? 0 : bytesToLong(appliedIndexBytes);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void destroy() {
+        // No-op.
     }
 
     private static void throwStorageStoppedException() {
         throw new IgniteInternalException(TX_STATE_STORAGE_STOPPED_ERR, "Transaction state storage is stopped");
     }
 
-    private byte[] uuidToBytes(UUID uuid) {
-        return ByteBuffer.allocate(2 * Long.BYTES).order(ByteOrder.BIG_ENDIAN)
-                .putLong(uuid.getMostSignificantBits())
-                .putLong(uuid.getLeastSignificantBits())
+    private byte[] txIdToKey(UUID txId) {
+        return ByteBuffer.allocate(Integer.BYTES + 2 * Long.BYTES).order(ByteOrder.BIG_ENDIAN)
+                .putInt(partitionId)
+                .putLong(txId.getMostSignificantBits())
+                .putLong(txId.getLeastSignificantBits())
                 .array();
     }
 
-    private UUID bytesToUuid(byte[] bytes) {
-        long msb = bytesToLong(bytes, 0);
-        long lsb = bytesToLong(bytes, Long.BYTES);
+    private UUID keyToTxId(byte[] bytes) {
+        long msb = bytesToLong(bytes, Integer.BYTES);
+        long lsb = bytesToLong(bytes, Integer.BYTES + Long.BYTES);
 
         return new UUID(msb, lsb);
     }
 
     /** {@inheritDoc} */
     @Override public void close() throws Exception {
-        stop();
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        List<AutoCloseable> resources = new ArrayList<>(iterators);
+
+        IgniteUtils.closeAll(resources);
+    }
+
+    /**
+     * Iterator for partition's transaction state storage that has shared underlying database.
+     */
+    private class StorageIterator implements Iterator<IgniteBiTuple<UUID, TxMeta>>, AutoCloseable {
```

Comment:
Why don't you use RocksIteratorAdapter? It was good enough in my implementation. Can you bring it back?